### PR TITLE
Moved the method that retracts all variables with a given delta to Objective

### PR DESCRIPTION
**Pull request by luisenp**
We have talked for quite a while about where to place this method. Proposing the solution in this PR. Planning to use this as the starting point for the vectorized retraction maps.

### CODE REVIEW

File: theseus/core/objective.py [hunk 1]
```diff
@@ -503,3 +503,20 @@ def to(self, *args, **kwargs):
         self.dtype = dtype or self.dtype
         if self._vectorization_to is not None:
             self._vectorization_to(*args, **kwargs)
+
+    def step_optim_vars(
```
Comment (mhmukadam) [inline theseus/core/objective.py:507]: Nice, Objective class actually seems like a good place for this function.

`step_...` makes it sound like optimizer related, how about `update_...`? Maybe `update_optim_vars_with_delta` if we want to be really descriptive.

File: theseus/optimizer/nonlinear/nonlinear_optimizer.py [hunk 2]
```diff
@@ -262,8 +262,11 @@ def _optimize_loop(
                 step_size = self.params.step_size
                 force_update = False
 
-            self.retract_and_update_variables(
-                delta, converged_indices, step_size, force_update=force_update
+            self.objective.step_optim_vars(
+                delta * step_size,
```
Comment (mhmukadam) [inline theseus/optimizer/nonlinear/nonlinear_optimizer.py:266]: Maybe worth adding an explicit line that does `delta *= step_size` before this function call?

File: theseus/theseus_layer.py [hunk 3]
```diff
@@ -218,8 +218,8 @@ def backward(ctx, *grad_outputs):
         with torch.no_grad():
             bwd_optimizer.linear_solver.linearization.linearize()
             delta = bwd_optimizer.linear_solver.solve()
-            bwd_optimizer.retract_and_update_variables(
-                delta, None, 1.0, force_update=True
+            bwd_optimizer.objective.step_optim_vars(
+                delta, bwd_optimizer.linear_solver.linearization.ordering
```
Comment (mhmukadam) [inline theseus/theseus_layer.py:222]: The previous version was doing `force_update=True` here?

**Review by mhmukadam — APPROVED** [review]

File: theseus/core/objective.py (hunk 1 above)
Comment (luisenp) [reply theseus/core/objective.py:507]: Maybe `retract_optim_vars`?

File: theseus/optimizer/nonlinear/nonlinear_optimizer.py (hunk 2 above)
Comment (luisenp) [reply theseus/optimizer/nonlinear/nonlinear_optimizer.py:266]: Is this necessary? I don't see delta being used anywhere else after this function call.

File: theseus/core/objective.py (hunk 1 above)
Comment (mhmukadam) [reply theseus/core/objective.py:507]: Yeah `retract_optim_vars` is good.

File: theseus/optimizer/nonlinear/nonlinear_optimizer.py (hunk 2 above)
Comment (mhmukadam) [reply theseus/optimizer/nonlinear/nonlinear_optimizer.py:266]: Probably okay, just wondering if there is a chance of introducing bugs in the future by having a variable change in a function call.

**Review by rtqichen — APPROVED** [review]

File: theseus/core/objective.py (hunk 1 above)
Comment (luisenp) [reply theseus/core/objective.py]: Changed.

**Review by mhmukadam — APPROVED** [review]

**Pull request merged**